### PR TITLE
fix: resolve square bracketed references

### LINKED ISSUE #145: resolve square bracketed references as xrefs
Currently, references of pattern `[name][link]` are not being handled. They also clutter the space a lot within the documentation and should be handled to link to the items as necessary.

### PR DESCRIPTION
**Pull request by dandhlee**
Square bracketed references of the form `[name][link]` will now properly be handled accommodating its regex format.

Towards b/195685296.
Fixes #145.

- [x] Tests pass

### CODE REVIEW

File: docfx_yaml/extension.py [hunk 1]
```diff
@@ -82,6 +82,7 @@ class Bcolors:
 REF_PATTERN = ':(py:)?(func|class|meth|mod|ref|attr|exc):`~?[a-zA-Z0-9_\.<> ]*(\(\))?`'
 # Regex expression for checking references of pattern like "~package_v1.subpackage.module"
 REF_PATTERN_LAST = '~([a-zA-Z0-9_<>]*\.)*[a-zA-Z0-9_<>]*(\(\))?'
+REF_PATTERN_BRACKETS = '\[[a-zA-Z0-9\_\<\>\-\. ]+\]\[[a-zA-Z0-9\_\<\>\-\. ]+\]'
```
Comment (tbpg) [inline docfx_yaml/extension.py:85]: Add a comment like for `REF_PATTERN_LAST`.

File: docfx_yaml/extension.py [hunk 2]
```diff
@@ -251,13 +252,18 @@ def _resolve_reference_in_module_summary(pattern, lines):
...
 
-            else:
+            elif pattern == REF_PATTERN_LAST:
                 index = matched_str.rfind('.') + 1
                 if index == 0:
                     # If there is no dot, push index to not include tilde
                     index = 1
                 ref_name = matched_str[index:]
 
+            else: #pattern == REF_PATTERN_BRACKETS:
```
Comment (tbpg) [inline docfx_yaml/extension.py:262]: Make this `elif` and `raise ValueError` in the `else` block to indicate a programming mistake.

File: docfx_yaml/extension.py [hunk 3]
```diff
@@ -818,17 +824,20 @@ def _update_friendly_package_name(path):
...
-        lines, xrefs = _resolve_reference_in_module_summary(REF_PATTERN_LAST, lines)
-        for xref in xrefs:
-            if xref not in app.env.docfx_xrefs:
-                app.env.docfx_xrefs[xref] = ''
+        ref_patterns = [
+            # REF_PATTERN checks for patterns like ":class:`~google.package.module`"
+            REF_PATTERN,
+            # REF_PATTERN_LAST checks for patterns like "~package.module"
+            REF_PATTERN_LAST,
+            # REF_PATTERN_BRACKETS checks for patterns like "~package.module"
```
Comment (tbpg) [inline docfx_yaml/extension.py:832]: Update comment.

File: docfx_yaml/extension.py [hunk 4]
```diff
@@ -251,13 +252,18 @@ def _resolve_reference_in_module_summary(pattern, lines):
                 # Find the last component of the target. "~Queue.get" only returns <xref:get>
                 ref_name = ref_name[index:]
 
-            else:
+            elif pattern == REF_PATTERN_LAST:
```
Comment (tbpg) [inline docfx_yaml/extension.py:255]: Consider (1) defining a function for each of these patterns and (2) calling all of the functions back to back with no for loop over patterns.

File: tests/test_unit.py [hunk 5]
```diff
@@ -234,7 +235,35 @@ def test_reference_in_summary_more_xrefs(self):
...
+            'google.cloud.kms.v1.KeyRing',
+            'google.cloud.kms.v1.ImportJob',
+        ]
+        summary_want = """Required.
+
+The <xref uid="google.cloud.kms.v1.KeyRing.name">name</xref> of the <xref uid="google.cloud.kms.v1.KeyRing">KeyRing</xref> associated with the <xref uid="google.cloud.kms.v1.ImportJob">ImportJobs</xref>.
+"""
+        summary_want = summary_want.split("\n")
+
+        xrefs_got = []
```
Comment (tbpg) [inline tests/test_unit.py:253]: Is this line needed?

File: tests/test_unit.py [hunk 6]
```diff
@@ -234,7 +235,35 @@ def test_reference_in_summary_more_xrefs(self):
...
+The [name][google.cloud.kms.v1.KeyRing.name] of the [KeyRing][google.cloud.kms.v1.KeyRing] associated with the [ImportJobs][google.cloud.kms.v1.ImportJob].
+"""
+        summary = summary.split("\n")
+
+        summary_got, xrefs = _resolve_reference_in_module_summary(REF_PATTERN_BRACKETS, summary)
+        for xref in xrefs:
+            xrefs_got.append(xref)
+
+        self.assertEqual(summary_got, summary_want)
+        self.assertCountEqual(xrefs_got, xrefs_want)
```
Comment (tbpg) [inline tests/test_unit.py:265]: Does the order of xrefs matter? `assertCountEqual` does not look at order. https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual

**Review by tbpg — CHANGES_REQUESTED** [review]

File: docfx_yaml/extension.py (hunk 4 above)
Comment (dandhlee) [reply docfx_yaml/extension.py:259]: I'm not exactly sure how I can achieve this with the given setup, I'll mark it as to-do for now and look back on this shortly when I refactor this codebase after launch.

**Review by dandhlee — COMMENTED** [review]
Thank you! Updated with comments and added a test case for the added exception.

File: docfx_yaml/extension.py [hunk 7]
```diff
@@ -252,13 +256,21 @@ def _resolve_reference_in_module_summary(pattern, lines):
...
                     index = 1
                 ref_name = matched_str[index:]
 
+            elif pattern == REF_PATTERN_BRACKETS:
+                lbracket = matched_str.find('[')+1
+                rbracket = matched_str.find(']')
+                ref_name = matched_str[lbracket:rbracket]
+
+            else:
+                raise ValueError('Encountered wrong ref pattern.')
```
Comment (tbpg) [inline docfx_yaml/extension.py:272]: Include the ref pattern in the exception message.

File: docfx_yaml/extension.py [hunk 8]
```diff
@@ -820,17 +832,21 @@ def _update_friendly_package_name(path):
...
-        for xref in xrefs:
-            if xref not in app.env.docfx_xrefs:
-                app.env.docfx_xrefs[xref] = ''
-
-        # REF_PATTERN_LAST checks for patterns like "~package.module"
-        lines, xrefs = _resolve_reference_in_module_summary(REF_PATTERN_LAST, lines)
-        for xref in xrefs:
-            if xref not in app.env.docfx_xrefs:
-                app.env.docfx_xrefs[xref] = ''
+        ref_patterns = [
```
Comment (tbpg) [inline docfx_yaml/extension.py:835]: Declare this next to the declarations of these vars?

**Review by tbpg — APPROVED** [review]
A couple more minor comments.

File: docfx_yaml/extension.py (hunk 7 above)
Comment (dandhlee) [reply docfx_yaml/extension.py]: Done.

File: docfx_yaml/extension.py (hunk 8 above)
Comment (dandhlee) [reply docfx_yaml/extension.py]: Done.

**Comment by dandhlee** [comment]
Filed https://github.com/googleapis/gapic-generator-python/issues/1086 to track issues for square bracketed references not being formed correctly. Spoke to @busunkim96 and @parthea about this offline. Merging as-is for now.

**Pull request merged**